File: app/src/main/java/posidon/launcher/wall/WallLoader.java
package posidon.launcher.wall;

import android.graphics.BitmapFactory;
import android.os.AsyncTask;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

class WallLoader extends AsyncTask<Void, Void, List<Wall>> {

    private final StringBuilder text = new StringBuilder();
    private AsyncTaskListener listener;

    @Override
    protected List<Wall> doInBackground(Void... params) {
        String repo = "https://raw.githubusercontent.com/leoxshn/walls/master/";
        try {
            String line;
            BufferedReader bufferReader = new BufferedReader(new InputStreamReader(new URL(repo + "index.json").openStream()));
            while ((line = bufferReader.readLine()) != null) text.append(line);
            bufferReader.close();

            JSONArray array = new JSONObject(text.toString()).getJSONArray("");
            List<Wall> walls = new ArrayList<>();
            int realI = 0;
            for (int i = 0; i < array.length(); i++) {
                try {
                    walls.add(realI, new Wall());
                    walls.get(realI).name = array.getJSONObject(i).getString("n");
                    walls.get(realI).author = array.getJSONObject(i).getString("a");
                    String d = repo + array.getJSONObject(i).getString("d");
                    walls.get(realI).img = BitmapFactory.decodeStream(new URL(d + "/thumb.jpg").openConnection().getInputStream());
                    walls.get(realI).url = d + "/img.png";
                    realI++;
                } catch (Exception ignore) {}
            }
            return walls;
        } catch (Exception e) { e.printStackTrace(); }
        return null;
    }

    void setListener(AsyncTaskListener listener) { this.listener = listener; }
    public interface AsyncTaskListener { void onAsyncTaskFinished(List<Wall> walls);}

    @Override
    protected void onPostExecute(List<Wall> walls) { if (listener != null) listener.onAsyncTaskFinished(walls); }
}

File: app/src/main/java/posidon/launcher/wall/WallLoader.kt
package posidon.launcher.wall

import android.graphics.BitmapFactory
import android.os.AsyncTask
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.URL

internal class WallLoader : AsyncTask<Unit, Unit, List<Wall>>() {

    private val text = StringBuilder()
    var listener: ((List<Wall>) -> Unit)? = null

    override fun doInBackground(vararg params: Unit): List<Wall> {
        val repo = "https://raw.githubusercontent.com/leoxshn/walls/master/"
        try {
            var line: String?
            val bufferReader = BufferedReader(InputStreamReader(URL(repo + "index.json").openStream()))
            while (bufferReader.readLine().also { line = it } != null) text.append(line)
            bufferReader.close()
            val array = JSONObject(text.toString()).getJSONArray("")
            val walls: ArrayList<Wall> = ArrayList()
            var realI = 0
            for (i in 0 until array.length()) {
                try {
                    walls.add(realI, Wall())
                    walls[realI].name = array.getJSONObject(i).getString("n")
                    walls[realI].author = array.getJSONObject(i).getString("a")
                    val d = repo + array.getJSONObject(i).getString("d")
                    walls[realI].img = BitmapFactory.decodeStream(URL("$d/thumb.jpg").openConnection().getInputStream())
                    walls[realI].url = "$d/img.png"
                    realI++
                } catch (ignore: Exception) {
                }
            }
            return walls
        } catch (e: Exception) { e.printStackTrace() }
        return emptyList()
    }

    override fun onPostExecute(walls: List<Wall>) { listener?.invoke(walls) }
}